### PR TITLE
TopEntityGeneration: fix for Vec's

### PR DESCRIPTION
**Pull request by blaxill**
- Fixed seen' bug
- Remove all Vec's from type tree during naming as:
  - Elements are type homogeneous so will generate the same name and
  - We don't try to provide number prefixes for elements
- Added Vec test

### CODE REVIEW

**Comment by blaxill** [comment]
`seen` was used instead of `seen'` in some recursive calls, which bypassed checking if the current name had been seen (it was a typo). This caused the erroneous cases to not terminate. With that fixed, recursive datatypes will still unroll once, so I removed them during the first pass. Another solution might be to backtrack during `gatherNames` when discovering a directly recursive type, but this would not be perfect either as circular referencing types wouldnt be caught. Unhandled cases would terminate under both solutions but may result in unexpected portnames when annotations exist inside the recursive types (which doesn't have clear semantics anyway.)

**Comment by blaxill** [comment]
That is to say

~~~Haskell
data Foo = Foo ("someName" :: Int)

Bar :: Vec _ Foo
==> Should now be e.g. PortName ""
~~~

but 

~~~Haskell
data Baz = Baz Foo Baz

Bar :: Baz
==> will be something like: PortProduct "" [PortName "someName"]
~~~

This is incorrect but a recursive type with annotation inside is either meaningless, or requires enumerating over to generate ordinal prefixes. I'm hesitant to do this because its going to get increasingly hairy to do in template haskell.

**Comment by christiaanb** [comment]
I guess a refactor to `ReaderT (Set.Set Name) Q` might be worth it, so you can't make this mistake in the future.

**Comment by blaxill** [comment]
Adding much broader / general parametrized type support. Types resolving to a single constructor after applying type parameters are the only candidates.

This brings the support of `makeTopEntity` closer to the support of clash compiler top entities. I'm going to make one more pass through this to clean stuff up a bit.

File: clash-prelude/src/Clash/Annotations/TH.hs [hunk 1]
```diff
@@ -284,15 +375,15 @@ expandFamilies t = embed <$> sequence t
 
 -- | Runs 'expandFamilies' and then 'gatherNames'
 expandFamiliesAndGatherNames
-  :: Set.Set Name
+  :: -- Set.Set Name
   -- ^ Locally track seen names to handle recursive datatypes
```
Comment (martijnbastiaan) [inline clash-prelude/src/Clash/Annotations/TH.hs]: This might very well be what's breaking Haddock on CI

**Comment by blaxill** [comment]
CI still looks fine so I'll remove remaining todos and squash

**Comment by christiaanb** [comment]
@blaxill Could you add a line to a changelog for this new feature? (not persé  for this fix for Vec, but for this extension to the port naming feature in general)

**Comment by blaxill** [comment]
> @blaxill Could you add a line to a changelog for this new feature? (not persé for this fix for Vec, but for this extension to the port naming feature in general)

Done and I've added a small summary of limitations for the feature too.

**Comment by blaxill** [comment]
@martijnbastiaan gitlab CI failure looks unrelated? I'm not sure why it's only failing for the last commit

**Comment by christiaanb** [comment]
@blaxill one of our dependencies is lacking an upper bound on `vty`, which has changed API

**Comment by christiaanb** [comment]
@blaxill just created #826, that will hopefully solve the issue once merged

**Comment by christiaanb** [comment]
Thanks!

**Pull request merged**